File: apps/java/src/test/java/chapter11/EvaluationDataTest.java
package chapter11;

import static chapter11.TribuoEvaluationTest.pick;
import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.within;
import static org.junit.jupiter.api.Assumptions.assumeTrue;

import chapter02.Table;
import chapter03.TribuoTrees;
import dataset.DataDir;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.List;
import java.util.Map;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.tribuo.classification.Label;
import org.tribuo.classification.evaluation.LabelEvaluation;
import org.tribuo.classification.evaluation.LabelEvaluator;
import support.StdoutCapture;

class EvaluationDataTest {
  private final Path survived = DataDir.dataDir().resolve("Survived.csv");
  private final Path cinema = DataDir.dataDir().resolve("cinema.csv");

  @BeforeEach
  void requireData() {
    assumeTrue(
        Files.exists(survived) && Files.exists(cinema),
        "学習データ Survived.csv・cinema.csv が配置されていない（gulp data:setup）");
  }

  private static void assertScores(
      Map<String, Double> expected, Map<String, Double> actual, double tolerance) {
    assertThat(actual.keySet()).containsExactlyInAnyOrderElementsOf(expected.keySet());
    expected.forEach(
        (name, value) -> assertThat(actual.get(name)).as(name).isCloseTo(value, within(tolerance)));
  }

  @Test
  @DisplayName("同じ分割なら Tribuo の評価器で採点した Survived の平均と一致する")
  void survivedSameAsTribuo() throws IOException {
    var data = Dataset.prepareSurvived(Table.load(survived));
    var folds = CrossValidation.kFold(data.x().size(), Experiments.N_SPLITS, Experiments.SEED);
    var positive = new Label("1");

    List<LabelEvaluation> evaluations =
        folds.stream()
            .map(
                fold -> {
                  var model =
                      TribuoTrees.train(
                          pick(data.x(), fold.train()), pick(data.t(), fold.train()), 2);
                  var test =
                      TribuoTrees.toDataset(
                          pick(data.x(), fold.test()), pick(data.t(), fold.test()));
                  return new LabelEvaluator().evaluate(model, test);
                })
            .toList();

    var scores =
        Experiments.evaluate(
            () -> new TribuoEvaluationTest.TribuoTree(2), data, Experiments.SURVIVED_METRICS);
    assertScores(
        Map.of(
            "正解率",
                evaluations.stream().mapToDouble(LabelEvaluation::accuracy).average().orElseThrow(),
            "適合率",
                evaluations.stream()
                    .mapToDouble(e -> e.precision(positive))
                    .average()
                    .orElseThrow(),
            "再現率",
                evaluations.stream().mapToDouble(e -> e.recall(positive)).average().orElseThrow(),
            "F値", evaluations.stream().mapToDouble(e -> e.f1(positive)).average().orElseThrow()),
        scores,
        1e-12);
  }

  @Test
  @DisplayName("実行すると交差検証の平均を表示する")
  void mainPrintsScores() throws Exception {
    String output = StdoutCapture.capture(() -> Main.main(new String[0]));

    assertThat(output)
        .isEqualTo(
            """
            Survived（決定木、5 分割交差検証の平均）
              正解率: 0.7811
              適合率: 0.7759
              再現率: 0.6306
              F値: 0.6833
            cinema（線形回帰、5 分割交差検証の平均）
              RMSE: 405.77
              MAE: 321.53
            """);
  }
}

File: apps/clojure/test/getting_started_ml/evaluation_data_test.clj
(ns getting-started-ml.evaluation-data-test
  "実データ（Survived.csv・cinema.csv）で交差検証するテスト。学習データが無ければスキップする。"
  (:require [clojure.string :as str]
            [clojure.test :refer [deftest is]]
            [getting-started-ml.chapter02 :as chapter02]
            [getting-started-ml.chapter11 :as ch]
            [getting-started-ml.dataset :as dataset])
  (:import [org.tribuo.classification Label]
           [org.tribuo.classification.evaluation LabelEvaluation LabelEvaluator]))

(defn- data-file
  "学習データのファイルを返す。無ければ理由を標準エラーに出して nil を返す。"
  [file-name]
  (let [path (str (dataset/dir) "/" file-name)]
    (if (.exists (java.io.File. path))
      path
      (binding [*out* *err*]
        (println (str "学習データ " file-name " が配置されていない（gulp data:setup）のでスキップする"))
        nil))))

(defn- survived-data
  "前処理した Survived の特徴量と正解ラベルを返す。学習データが無ければ nil。"
  []
  (when-let [path (data-file "Survived.csv")]
    (ch/prepare-survived (chapter02/load-table path))))

(defn- mean-of
  "分割ごとの評価結果を、採点する関数で平均する。"
  [evaluations score]
  (ch/mean (mapv score evaluations)))

(deftest 実データを八百九十一件読み欠損値の年齢を平均値で補う
  (when-let [{:keys [x t]} (survived-data)]
    (is (= 891 (count x)))
    (is (= 891 (count t)))
    (is (every? (fn [features] (every? some? (vals features))) x))
    (is (= #{"0" "1"} (set t)))))

(deftest 同じ分割ならTribuoの評価器で採点したSurvivedの平均と一致する
  (when-let [{:keys [x t]} (survived-data)]
    (let [folds (ch/k-fold (count x) ch/n-splits ch/seed)
          positive (Label. "1")
          evaluations (mapv (fn [{:keys [train test]}]
                              (let [model (ch/tribuo-tree (ch/pick x train) (ch/pick t train)
                                                          ch/survived-columns 2)]
                                (.evaluate (LabelEvaluator.) model
                                           (ch/to-dataset (ch/pick x test) (ch/pick t test)
                                                          ch/survived-columns))))
                            folds)
          scores (into {} (ch/evaluate (ch/tribuo-tree-trainer ch/survived-columns 2)
                                       {:x x :t t} ch/survived-metrics))]
      (is (< (abs (- (mean-of evaluations #(.accuracy ^LabelEvaluation %)) (get scores "正解率")))
             1e-12))
      (is (< (abs (- (mean-of evaluations #(.precision ^LabelEvaluation % positive))
                     (get scores "適合率")))
             1e-12))
      (is (< (abs (- (mean-of evaluations #(.recall ^LabelEvaluation % positive))
                     (get scores "再現率")))
             1e-12))
      (is (< (abs (- (mean-of evaluations #(.f1 ^LabelEvaluation % positive)) (get scores "F値")))
             1e-12)))))

(deftest 実行すると交差検証の平均を表示する
  (when (and (data-file "Survived.csv") (data-file "cinema.csv"))
    (is (= ["Survived（決定木、5 分割交差検証の平均）"
            "  正解率: 0.7811"
            "  適合率: 0.7759"
            "  再現率: 0.6306"
            "  F値: 0.6833"
            "cinema（線形回帰、5 分割交差検証の平均）"
            "  RMSE: 405.77"
            "  MAE: 321.53"]
           (str/split-lines (with-out-str (ch/run)))))))
